### PR TITLE
Gather timestamps during the execution of the request

### PR DESCRIPTION
**Pull request by Rumbeer**
Added gathering of timestamps during the exectuion of the request described in https://github.com/riganti/dotvvm/issues/369. It uses the IStopwatch interface to register a System.Diagnostics.Stopwatch singleton object.

### CODE REVIEW

File: src/DotVVM.Framework/Hosting/DotvvmPresenter.cs [hunk 1]
```diff
@@ -249,6 +262,14 @@ public class DotvvmPresenter : IDotvvmPresenter
             }
         }
 
+        private long addTraceData(long lastStopwatchState, string processName, IDotvvmRequestContext context)
```
Comment (tomasherceg) [inline src/DotVVM.Framework/Hosting/DotvvmPresenter.cs]: Please rename the method to `AddTraceData`.
And the rename the `processName` to `eventName`.

File: src/DotVVM.Framework/Hosting/DotvvmPresenter.cs [hunk 2]
```diff
@@ -249,6 +262,14 @@ public class DotvvmPresenter : IDotvvmPresenter
...
+        {
+            var stopwatch = context.Configuration.ServiceLocator.GetService<IStopwatch>();
```
Comment (tomasherceg) [inline src/DotVVM.Framework/Hosting/DotvvmPresenter.cs]: We should get the stopwatch instance in the constructor so we do not call `GetService` for every traced event.

File: src/DotVVM.Framework/Hosting/DotvvmRequestContext.cs [hunk 3]
```diff
@@ -77,6 +77,11 @@ public class DotvvmRequestContext : IDotvvmRequestContext
         /// </summary>
         public ModelState ModelState { get; private set; }
 
+        /// <summary>
+        /// Gets the data gathered during execution of the request.
+        /// </summary>
+        public Dictionary<string, object> TraceData { get; set; }
```
Comment (tomasherceg) [inline src/DotVVM.Framework/Hosting/DotvvmRequestContext.cs]: Make the setter private.

**Comment by Rumbeer** [comment]
Added commit with requested changes (AddTraceData gets instance of IStopwatch through constructor, rename addTraceData to AddTraceData and processName to eventName, the TraceData setter is now private)

**Comment by Rumbeer** [comment]
Added constants containing event names used during exection of a request.

**Comment by exyi** [comment]
Thank you for the PR, I like this feature but there are few things I’d do a bit differently:
• I don’t understand what is the purpose of IStopwatch interface. Do you think it would be a good idea to implement custom clock?
• Instead, I’d like to inject a tracer, that would collect these events and log them somewhere.
• I don’t like the another property on an already bulky IDotvvmRequestContext. What about to make a request scoped service or a property on HttpRequest (in the Items dictionary)?
• Why is the event id a string? IMO it would be much better to have an enum for them.

**Comment by tomasherceg** [comment]
That's mainly my fault because I worked with @Rumbeer on this.

1) I wanted an abstraction over Stopwatch,. mainly because registering Stopwatch instance in the service locator is not the best idea (any other framework can do the same and they may need another lifetime of the instance which would break us).

2) That's solved in another issue, we are doing the Application Insights tracer. https://github.com/riganti/dotvvm/issues/370 
The purpose of this is only to gather data for the further processing. The tracer will be able to get this collection and do whatever it wants with it. And you'll be able to have multiple tracers.

3) I don't like it either but we should review everything that is on the context and decide what to keep and what to move elsewhere. I am sure we will find more properties that should not be there. 

4) I'd like to use this collection for more things in the future. The enum would be changed very soon. 
And the application developers may want to add their own data in this collection, so the tracer can attach them to the HTTP request information. 

I have merged this PR so we can continue working on the App Insights integration.

**Pull request merged**